### ISSUE TITLE
Meshnet questions

### ISSUE THREAD

**Issue by benhylau** [dweb-camp-2019/meshnet#2]
Moved from https://github.com/dweb-camp-2019/organizing/issues/17

This issue collects a list of questions that the prototyping process needs to answer.

_The top message will be updated as more questions emerge that I don't currently have an answer for. Please add questions to the thread._

---

- [x] Will/should I be able to host a service on the network from my laptop that is connected to the network? Or will I need to bring a microcomputer (e.g. rpi) and plug into one of these official planned nodes? https://github.com/dweb-camp-2019/meshnet/issues/2#issuecomment-496879418

- [x] Will their be an on-site DNS server that's consistently managed for internal FQDNs? Or, will there be extensive use of mDNS implementations like Avahi? https://github.com/dweb-camp-2019/meshnet/issues/2#issuecomment-501932980

- [x] Will multicast UDP be supported? If so, it could be an easy way to do service discovery across the network using [Multicast DNS](https://en.wikipedia.org/wiki/Multicast_DNS) https://github.com/dweb-camp-2019/meshnet/issues/2#issuecomment-501932980

- [x] How do I provide a web service to attendees via the mesh network? https://github.com/dweb-camp-2019/meshnet/issues/2#issuecomment-496879418

- [x] What kind of work can I do ahead of time so minimize amount of hacking I need to do at the event? (i.e. do I need to bring my own device? Does that device need to have an Ethernet port or WiFi radio? Will there be 'golden servers' that are preconnected and I should bring a USB driver with software in a certain format like a docker image or tarball?) https://github.com/dweb-camp-2019/meshnet/issues/2#issuecomment-496879418

**Comment by raylas** [comment]
- [ ] Will their be an on-site DNS server that's consistently managed for internal FQDNs? Or, will there be extensive use of mDNS implementations like Avahi?

**Comment by RangerMauve** [comment]
[ ] Will multicast UDP be supported? If so, it could be an easy way to do service discovery across the network using [Multicast DNS](https://en.wikipedia.org/wiki/Multicast_DNS)

**Comment by amark** [comment]
@RangerMauve will you be at the camp!? We'll finally get to meet in person, after ?5? years!

Yeah, we have some multicast stuff already experimented, this would be a great place to try it in action (or bluetooth discovery). I 2nd multicast!

**Comment by RangerMauve** [comment]
@amark Ha ha, hopefully! It sounds like a lot of cool people are going to be attending. :D

**Comment by gobengo** [comment]
We're about six weeks from camp time. When/how will these questions get answered?

In particular I'm curious about DNS exactly what the process would be to host something like an HTTP Server on the mesh with an address I can tell other people about. Ideally they can then access that service using their phones without having to configure their phone os networking in any fancy ways. And double ideally there is some kind of reasonable DNS name I can tell them without having to write down an IP address or other hard-to-remember name.

In short
* [ ] How do I provide a web service to attendees via the mesh network?
* [ ] What kind of work can I do ahead of time so minimize amount of hacking I need to do at the event? (i.e. do I need to bring my own device? Does that device need to have an Ethernet port or WiFi radio? Will there be 'golden servers' that are preconnected and I should bring a USB driver with software in a certain format like a docker image or tarball?)

**Comment by benhylau** [comment]
We are just picking up on this and most of the conversations have been on the riot chat, so thanks for pushing me to add documentation on this. I hope this answers your questions: https://github.com/dweb-camp-2019/meshnet/tree/master/testnet

>How do I provide a web service to attendees via the mesh network?

Plug in a device, like `A` and `D` in the doc above, you will be assigned a locally-unique and mesh-public IP. That still leaves the DNS question unanswered, we'll move to https://github.com/dweb-camp-2019/meshnet/issues/12 to discuss that, and perhaps some high bandwidth chat [on matrix](https://riot.im/app/#/room/#dweb-camp-2019-mesh:matrix.org) would be useful too.

>What kind of work can I do ahead of time so minimize amount of hacking I need to do at the event?

The testnet setup would be a good emulation of the event meshnet, it requires two Raspberry Pi 3B / 3B+. (I am pretty sure a Pi 2 with compatible WiFi adapter works too, but didn't test.) Aside from that, we are planning June 26 to be a test date at IA, with a reduced version of the meshnet there but with actual MicroTik equipment, and we can gather and test all day. More details on that soon!

>do I need to bring my own device?

Yes please bring your own device with ethernet port. WiFi should work too but wired connection is always more reliable, as I am worried the area where we put devices is the same area people are hacking and crowding the local WiFi spectrum.

I think BYOD is simplest, but closer to event date maybe we can plan a server with docker, but I am afraid the IP assignment may get more convoluted.

**Comment by benhylau** [comment]
@gobengo oh yea we also have a bunch of Raspberry Pis available from last Summit, you can bring an SD to plug in if you like, just let me know ahead of time!

**Comment by mitra42** [comment]
If you could reserve one of those RPIs for me would be great, I can bring an SD or we can install on it live (meaning it could have any meshing/networking setup already and I'll install Internet In A Box with the Internet Archive server built in.

**Comment by gobengo** [comment]
@benhylau Thanks for the responses. They are very helpful.

**Comment by amark** [comment]
@mitra42 we've rolled out quite a few significant improvements (multicast support, etc.) and am rolling out significant scaling improvements (we had almost 10,000 real world users connected using ~90MB RAM), routing, etc. (not published yet, just on our test system), will there be a time & place for to upgrade IA's dWeb GUN version before camp?

**Comment by mitra42** [comment]
Maybe - but thats a question for a different place than the meshnet conversation.

**Comment by mitra42** [comment]
@amark email sent.

**Comment by benhylau** [comment]
This is the network topology currently set up at the Internet Archive. As more of the equipment arrives we'll have 3 nodes set up here, in time for the June 26 _Project prototyping day at Internet Archive_. I wrote a little about the setup [on Twitter](https://twitter.com/LauBenedict/status/1138628795693277184) but will write a proper blog post once I have the vlan stuff sorted and MikroTik scripts published.

![IMG_20190612_103832_1](https://user-images.githubusercontent.com/2261308/59476305-00bedd80-8e05-11e9-981d-975622e95ad0.jpg)

Raspberry Pis or other application-serving devices will connect to a PoE switch like this, and we will buy a bunch of 5V PoE splitters. Of course, please bring if you have some:

![IMG_20190613_121708](https://user-images.githubusercontent.com/2261308/59476361-2f3cb880-8e05-11e9-93aa-619e3b9b79fb.jpg)

The indoor [Hacking Room](https://dwebcamp.org/proposals-spaces/) will have a 24-port 10/100 PoE switch. Wireless devices are on same LAN as indicated by first graphic.

>Will their be an on-site DNS server that's consistently managed for internal FQDNs? Or, will there be extensive use of mDNS implementations like Avahi? --@raylas

I hope https://github.com/dweb-camp-2019/meshnet/issues/12#issuecomment-501920788 answers your question. Basically:

- Manual _Directory Listing_ page at `10.0.0.1` that maps IP address to description
- mDNS over LAN on same node only
- Internet DNS over DHCP
- Encouraging people to run local DNS servers

>Will multicast UDP be supported? If so, it could be an easy way to do service discovery across the network using [Multicast DNS](https://en.wikipedia.org/wiki/Multicast_DNS) --@RangerMauve 

Same node only. Across nodes, SSB will need Pub servers, IPFS would need manual bootstrap. But addresses are unique and routable across all nodes (i.e. no NAT), so at least it's easy to set up those bootstrap servers.

Sorry I took a long time to answer these. How do you guys feel about this setup? Also any general thoughts about unplugging the Internet backhaul for scheduled hours of the day?

**Comment by mitra42** [comment]
@benhylau I have a RPI in the cable closet - Jona knows where it is, its currently on rpi.sf.archive.org but could easily be moved to the mesh. If you are going to do it, then give me ~24-48 hours notice to do an upgrade to current version of IIAB / dweb-mirror first
